groups/ntc/ntcs/ntcs_proactordetachcontext.h

```diff
@@ -19,49 +19,370 @@
+    bool                                         d_authorization;
 
   private:
-    bsls::AtomicUint d_stateAndCounter;
+    ProactorDetachGuard(const ProactorDetachGuard&) BSLS_KEYWORD_DELETED;
+    ProactorDetachGuard& operator=(const ProactorDetachGuard&) 
+        BSLS_KEYWORD_DELETED;
+
+  public:
+    /// Create a new guard, detect if operation on the specified 'socket' is
+    /// authorized, and if so, automatically increment the number of processors
+    /// of the 'socket'. Note that authorization is indicated by evaluating
+    /// the resulting object as a boolean.
+    explicit ProactorDetachGuard(
+        const bsl::shared_ptr<ntci::ProactorSocket>& socket);
+    
+    /// Unless otherwise released, decrement the number of processors of the
+    /// underlying socket if the guard authorized the caller to operate of the
+    /// socket, schedule the completion of the asynchronous socket detach
+    /// operation, if appropriate, and destroy this object.
+    ~ProactorDetachGuard();
+
+    /// Release the guard from managing the underlying socket.
+    void release();
 
-    enum { k_STATE_MASK = 0xC0000000 };
-    enum { k_COUNTER_MASK = 0x3FFFFFF };
+    /// Return true if the guard authorizes the calling thread to actively work
+    /// on the socket, otherwise return false.
+    operator bool() const;
 };
 
+NTCCFG_INLINE
+bool ProactorDetachContext::update(unsigned int oldValue, 
+                                   unsigned int newValue)
+{
+    return d_value.testAndSwap(oldValue, newValue) == oldValue;
+}
+
+NTCCFG_INLINE
+ProactorDetachContext::ProactorDetachContext()
+: d_value(0)
+{
+}
+
+NTCCFG_INLINE
+ProactorDetachContext::~ProactorDetachContext()
+{
+}
+
+NTCCFG_INLINE
+bool ProactorDetachContext::incrementReference()
+{
+    while (true) {
+        unsigned int currentValue = d_value.load();
+        unsigned int currentState = currentValue & k_STATE_MASK;
+        unsigned int currentCount = currentValue & k_COUNT_MASK;
+
+        if (currentState == k_ATTACHED) {
+            if (update(currentValue, k_ATTACHED | (currentCount + 1))) {
+                return true;
+            }
+        }
+        else {
+            return false;
+        }
+    }
+}
+
+NTCCFG_INLINE
+bool ProactorDetachContext::decrementReference()
+{
+    while (true) {
+        const unsigned int currentValue = d_value.load();
+        const unsigned int currentState = currentValue & k_STATE_MASK;
+        const unsigned int currentCount = currentValue & k_COUNT_MASK;
+
+        if (currentState == k_ATTACHED) {
+            BSLS_ASSERT(currentCount > 0);
+            if (update(currentValue, k_ATTACHED | (currentCount - 1))) {
+                return false;
+            }
+        }
+        else if (currentState == k_DETACHING) {
+            if (currentCount == 1) {
+                if (update(currentValue, k_DETACHED)) {
+                    return true;
+                }
+            }
```

Comment:
I think we should keep in mind that it's generally better to have less `if` statements so what about rewriting this part as:
```
BSLS_ASSERT(currentCount >= 1);
if (update(currentValue, k_DETACHING | (currentCount - 1))) {
    return (currentValue == 0);
}
```

Reply:
I interpret this suggestion as a desire to collapse some parts of the if-else-if-else blocks. We still need to conditionally transition to k_DETACHING or k_DETACHED, and it's not clear from your suggestion how you'd like to see this collapsed. We could rewrite this as a ternary expression but the branches are still there and I think the readability of the algorithm is worse.